torch/_tensor.py

```diff
@@ -1033,6 +1034,66 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
             else:
                 return _convert(ret, cls)
 
+    def __dlpack__(self, stream=None):
+        """
+        Creates a DLpack `capsule https://data-apis.org/array-api/latest/design_topics/data_interchange.html#data-interchange`_
+        of the current tensor to be exported to other libraries.
+
+        This function will be called from the `from_dlpack` method
+        of the library that will consume the capsule. `from_dlpack` passes the current
+        stream to this method as part of the specification.
+
+        Args:
+            stream (integer or None): A Python integer representing a pointer
+            to a stream (CUDA or ROCm). `stream` is provided by the consumer
+            to the producer to instruct the producer to ensure that operations
+            can safely be performed on the array.
+            The pointer must be a positive integer or
+            -1 . If stream is -1 , the value may be used by the consumer to
+            signal "producer must not perform any synchronization. Optional.
+        """
+        if has_torch_function_unary(self):
+            return handle_torch_function(Tensor.__dlpack__, (self,), self, stream)
+
+        # Some semantics that can prevent tensors from being exported are
```

Comment:
More precisely, DLPack capsules can't capture all of PyTorch's semantics, so we prohibit exporting tensors that would lose their properties like requires_grad and having the conjugate bit set.